File: WMS.Application/Services/AttendanceService.cs
using AutoMapper;
using WMS.Application.DTOs.Attendance;
using WMS.Application.Interfaces;
using WMS.Domain.Entities;

namespace WMS.Application.Services;

public class AttendanceService : IAttendanceService
{
    private readonly IAttendanceRepository _repo;
    private readonly IMapper _mapper;

    public AttendanceService(IAttendanceRepository repo, IMapper mapper)
    {
        _repo = repo;
        _mapper = mapper;
    }

    public async Task<IEnumerable<AttendanceDto>> GetAllAsync()
        => _mapper.Map<IEnumerable<AttendanceDto>>(await _repo.GetAllAsync());

    public async Task<AttendanceDto?> GetByIdAsync(int id)
    {
        var a = await _repo.GetByIdAsync(id);
        return a == null ? null : _mapper.Map<AttendanceDto>(a);
    }

    public async Task<IEnumerable<AttendanceDto>> GetByEmployeeAsync(int empId)
        => _mapper.Map<IEnumerable<AttendanceDto>>(await _repo.GetByEmployeeAsync(empId));

    public async Task<IEnumerable<AttendanceDto>> GetMonthlyAsync(int empId, int year, int month)
        => _mapper.Map<IEnumerable<AttendanceDto>>(await _repo.GetMonthlyAsync(empId, year, month));

    public async Task<int> CheckInAsync(CheckInDto dto)
    {
        var today = DateTime.Today;
        if (await _repo.HasAttendanceTodayAsync(dto.EmpId, today))
            throw new InvalidOperationException("Employee has already checked in today.");

        var attendance = new Attendance
        {
            EmpId = dto.EmpId,
            CheckIn = DateTime.Now,
            AttendanceDate = today,
            WorkMode = dto.WorkMode
        };
        await _repo.AddAsync(attendance);
        await _repo.SaveChangesAsync();
        return attendance.AttendanceId;
    }

    public async Task<bool> CheckOutAsync(int attendanceId, CheckOutDto dto)
    {
        var attendance = await _repo.GetByIdAsync(attendanceId);
        if (attendance == null || attendance.CheckOut.HasValue) return false;
        
        attendance.CheckOut = DateTime.Now;
        attendance.TotalHours = Math.Max(0, (attendance.CheckOut.Value - attendance.CheckIn).TotalHours);
        
        _repo.Update(attendance);
        await _repo.SaveChangesAsync();
        return true;
    }

    public async Task<int> CreateAsync(CreateAttendanceDto dto)
    {
        var attendance = _mapper.Map<Attendance>(dto);
        if (attendance.CheckOut.HasValue)
            attendance.TotalHours = (attendance.CheckOut.Value - attendance.CheckIn).TotalHours;
        await _repo.AddAsync(attendance);
        await _repo.SaveChangesAsync();
        return attendance.AttendanceId;
    }

    public async Task<bool> UpdateAsync(UpdateAttendanceDto dto)
    {
        var attendance = await _repo.GetByIdAsync(dto.AttendanceId);
        if (attendance == null) return false;
        _mapper.Map(dto, attendance);
        if (attendance.CheckOut.HasValue)
            attendance.TotalHours = (attendance.CheckOut.Value - attendance.CheckIn).TotalHours;
        _repo.Update(attendance);
        await _repo.SaveChangesAsync();
        return true;
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var a = await _repo.GetByIdAsync(id);
        if (a == null) return false;
        _repo.Delete(a);
        await _repo.SaveChangesAsync();
        return true;
    }

    public async Task<byte[]> GenerateTimesheetCsvAsync(int year, int month)
    {
        var all = await _repo.GetAllAsync();
        var monthly = all.Where(a => a.AttendanceDate.Year == year && a.AttendanceDate.Month == month).ToList();

        var sb = new System.Text.StringBuilder();
        sb.AppendLine("AttendanceId,EmpId,AttendanceDate,CheckIn,CheckOut,TotalHours,WorkMode");
        foreach (var a in monthly)
        {
            sb.AppendLine($"{a.AttendanceId},{a.EmpId},{a.AttendanceDate:yyyy-MM-dd},{a.CheckIn:HH:mm},{a.CheckOut?.ToString("HH:mm") ?? ""},{a.TotalHours},{a.WorkMode}");
        }
        return System.Text.Encoding.UTF8.GetBytes(sb.ToString());
    }
}

File: WMS.Frontend/src/app/shared/services/attendance.service.ts
import { environment } from '../../../environments/environment';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Attendance, CreateAttendanceDto } from '../models/models';

export interface CheckInDto { empId: number; checkIn: string; workMode: string; attendanceDate: string; }
export interface CheckOutDto { checkOut: string; }

@Injectable({ providedIn: 'root' })
export class AttendanceService {
  private base = '${environment.apiUrl}/attendance';
  constructor(private http: HttpClient) {}
  getAll() { return this.http.get<Attendance[]>(this.base); }
  getById(id: number) { return this.http.get<Attendance>(`${this.base}/${id}`); }
  getByEmployee(empId: number) { return this.http.get<Attendance[]>(`${this.base}/employee/${empId}`); }
  getMonthly(empId: number, year: number, month: number) {
    return this.http.get<Attendance[]>(`${this.base}/monthly/${empId}/${year}/${month}`);
  }
  exportTimesheet(year: number, month: number) {
    return this.http.get(`${this.base}/timesheet/export/${year}/${month}`, { responseType: 'blob' });
  }
  checkIn(dto: CheckInDto) { return this.http.post<any>(`${this.base}/checkin`, dto); }
  checkOut(id: number, dto: CheckOutDto) { return this.http.put(`${this.base}/checkout/${id}`, dto); }
  create(dto: CreateAttendanceDto) { return this.http.post<number>(this.base, dto); }
  update(dto: any) { return this.http.put(this.base, dto); }
  delete(id: number) { return this.http.delete(`${this.base}/${id}`); }
}


